### ISSUE TITLE
Should the server decide how to render forms instead of the client?

### ISSUE THREAD

**Issue by alexweissman** [userfrosting/UserFrosting#334]
Right now, the "forms" API lets the client specify a list of fields, and for each field, specify a render mode (`show`,`hidden`,`disabled`,`readonly`).  Usually these are parameters that are passed in through an AJAX call, requesting for example, the "User" form.  

The server responds by generating the form with the appropriate fields and in the requested modes.  Of course, if the client doesn't have permission to actually edit one of these fields (for example, "email"), then they will get rejected when they actually try to submit the form.  Nevertheless, they will be presented with an editable "email" field if they ask for it.

Should we instead just automatically check which fields the requesting user has access to, and set the render mode accordingly?  For example, we could first check to see if the requesting user can edit the "email" field for a particular user, and then set it to "disabled" or not depending on that setting.

**Comment by chrison999** [comment]
If the user doesn't have permission to edit a field then they shouldn't be presented with an editable field and then get rejected when they submit the form.  I vote for checking the user's permission before show generating the form, and then generating the form based on the user's permissions.

One "gotcha" to this...  the access control system should have the capability of distinguishing between "show but don't allow edits" and "don't show."  There might be cases where I want the user to see what's in a field but not allow them to edit it, and other cases where I don't even want them to see it because of their permissions.

Regards,

Chris

**Comment by alexweissman** [comment]
Awesome, thanks for the feedback.  So, I think we could do something like this:

```
foreach field in form:
    if (user->hasPermission("edit", field)):
        display[field] = "show"
    else if (user->hasPermission("view", field)):
        display[field] = "disabled"
    else:
        display[field] = "hidden"
```

We already have the functionality to assign and check permissions on that level, so this should be pretty straightforward.  However, it is worth noting that each call to `hasPermission` requires a database query.  So in forms with lots of fields, this could slow things down a bit.  To deal with this, maybe developers could come up with higher-level authorization tokens like "editShippingAddress", "editProfile", etc.

**Comment by ConnorVG** [comment]
You can load in the user's permissions and then check against that loaded data rather than re-quering? Not sure why you'd want to keep re-querying.

**Comment by alexweissman** [comment]
Well the problem is that permissions are not always necessarily as simple as a list of allowed fields.  For example, some users might only be able to access a field if some other conditions are met (time, links to other objects, the value of the field itself, etc). 

So, it would be hard to directly look up in the database and say "these are the fields the user has permission to edit" with one query.

**Comment by alexweissman** [comment]
This was implemented a long time ago in 0.3.1 - not sure why it's still open!
